### ISSUE TITLE
Support Cassandra in PetSet

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#24030]
__I haven't deployed Cassandra to production, I'll incorporate comments from anyone who has into this missive.__

A cassandra cluster is a ring of nodes. 
* Clients send request to any node (coordinator) which forwards requests to the node with the data.
* Each node is assigned a token. Each row is hashed based on a partition function applied on its primary key. A node stores hashes between its token and the next node's token. 
* A replication strategy picks how data is replicated, eg: 'SimpleStrategy' picks the next N nodes in the ring. The coordinator is also responsible for proxying writes to these nodes based on the consistency level of the request.

## Initial Deployment

To deploy a cassandra cluster you need to pick some "seed" nodes. These are nodes that come up first, and that all other nodes contact to get information about the cluster. They're only used in bootstrapping, after which all nodes cache their peers. All nodes in a cluster must have the same seed. Without at least one seed node, no nodes can join the cluster.

There appear to be many common ways to pick a seed:
See http://docs.datastax.com/en/archived/cassandra/1.0/docs/cluster_architecture/gossip.html.
* Priam uses Amazon's simpleDB to store each node's id as it comes up, and allocates the first in each DC as a seed (https://github.com/Netflix/Priam/wiki/Setup#s3-bucket-and-simpledb-domains). This requires centralized storage that isn't attached to any pet.
* Our [example](https://github.com/kubernetes/kubernetes/tree/master/examples/cassandra) uses Service resolution thorugh a custom entrypoint. If a node finds no endpoints in the Service, it inserts it's own IP into the seed list (TODO: follow up, this feels racy).

Perils:
* Setting all nodes as seeds
* Setting different seeds for different nodes in a single DC 
* Bringing up non-seeds before seeds

Kube implementation notes:    
* Calculate tokens for all nodes (eg: http://docs.datastax.com/en/cassandra/2.1/cassandra/configuration/configGenTokens_c.html)
* First node that comes up is seed, using hostnames is OK (risky to set 2 nodes as seeds that end up waiting on each other, TODO: there are probably better ways)
* Every subsequent node is allocated a segment of the token ring, and told about the seed.

## Scaling

Scaling seems straightforward, there are 3 common ways:
* Double size: Existing nodes can keep their assigned tokens, the new nodes just bisect or trisect the ring.
* Add one node: Leave its inital_token field empty, Cassandra will figure out the heaviest node token and split it.
* Add a random number of nodes: This involves recalculating tokens for the entire cluster and should probably be avoided.

Prefer GossipingSnitch to PropertyFileSnitch, the former requires updating each node with updated DC/rack information with new nodes.

### Adding a DC

The seed list of all nodes should include one node from each DC, or they will be unaware of the entire topology. This might require a total babysitter, and some way to update existing DCs with new seeds. A LOCAL_QUORUM setting is also recommended so we don't round trip to all DCs for every write.

## Failure

Cassandra seems to take a while to detect node failure because it uses gossip. Each node contacts some others periodically, to exchange timestamped heartbeats. These gossip heartbeats could take a while to propogate (or a lack thereof, a while to get detected). The removal of dead nodes is slightly tricky, because of the segmented keyspace and the gossip protocol:
* Confirm node is dead by running nodetool ring
* Set inital_token to dead nodes token (visible in `nodetool ring` output)
* Run nodetool repair on each keyspace of the new node
* Remove the dead node

### Replacing a dead seed

* Promote an existing node to seed
* Remove the dead seed from cassandra.yaml for each node in cluster
* Confirm that all nodes list new node as seed, old node as dead
* Perform steps above to replace a failed node

If a new node is integrated into the cluster before all nodes realize the old node is dead, or see the new seed if the dead node was a seed, the new node might start streaming data from the dead node. The good news is Cassandra will only fail a write if there are too few replicas alive, i.e if it sends a write to a dead node that just hasn't been observed by the cluster, the write will timeout and Cassandra will send the write elsewhere.

## Maintenance and Upgrade

One needs to run nodetool repair periodically on all nodes, within `gc_grace_seconds`. This will cause a spike in cpu. Running on multiple nodes simultaneously might affect throughput. It's also wise to backup SSTables snapshots to cloud storage, since they're immutable.

Upgrade is more tricky. Backup cluster scheme and ring tokens, then on each node:

- Create snapshot (nodetool snapshot)
- Drain node (nodetool drain)
- Stop cassandra
- Purge cassandra/saved_caches. TODO: Is is safer to wait for all nodes to realize this node is down?
- Start new cassandra

## Thoughts

Easier:
* Doubling size of cluster
* Bootstrapping, once you have the seeds up

Harder:
* Deleting nodes, scaling down
* Scaling up to some number of nodes that requires token recomputation
* Upgrading

Periodic backup of sstable snapshots is also recommended by many.
xref: https://github.com/kubernetes/kubernetes/issues/20961#issuecomment-204485326

**Comment by bprashanth** [comment]
@chrislovecnm

**Comment by chrislovecnm** [comment]
I have a few comments on this.  No time to comment right now :)  One feature that is missing is the capability of K8 to run a command, before it removes a node and a docker. i.e. `nodetool drain`

BTW vnodes make your life ALOT easier to scale up.  Will comment when I have time :)

**Comment by chrislovecnm** [comment]
Our example uses Service resolution thorugh a custom entrypoint. If a node finds no endpoints in the Service, it inserts it's own IP into the seed list (TODO: follow up, this feels racy).
Perils:

Setting all nodes as seeds

- no bad practice, bootstrapping a seed node is different. Recovery is different. Couple of other reasons to NOT do this.

Setting different seeds for different nodes in a single DC

- yep
Bringing up non-seeds before seeds

- again - run for the hills. Say set x > 3 nodes as seeds.

**Comment by chrislovecnm** [comment]
Failure section:

Set inital_token to dead nodes token (visible in nodetool ring output) 

- this is different with vnodes - I am only recommending vnodes as best practice with kubernetes

**Comment by chrislovecnm** [comment]
Maint and upgrade section: this depends on a couple of things:

- can we put the data on the k8 node and deploy a new docker? If so then you don't need to drain. If the data is going to be stored in the docker, then yes you need to drain
- upgrade process depends on version jumps as well.
- does k8 have a job scheduler on the roadmap? This could run nodetool repair.

**Comment by chrislovecnm** [comment]
Thoughts

Easier:

Doubling size of cluster. 

- use vnodes, you can add 1 or 1000 - does not matter

Bootstrapping, once you have the seeds up
Harder: 

- not sure what you mean

Deleting nodes, scaling down

- a hook - run this command before you stop this docker, would fix this 

Scaling up to some number of nodes that requires token recomputation

- not with vnodes ;)

**Comment by chrislovecnm** [comment]
Our example uses Service resolution thorugh a custom entrypoint. If a node finds no endpoints in the Service, it inserts it's own IP into the seed list (TODO: follow up, this feels racy).


I like this and don't like this ;). The k8 seed provider needs to be documented, and I am wondering if we can recommend a supported provider - ec2, GCE - etc

At the min it needs to be documented better :P. Also the damn thing is not compiled on the same version of C* as the docker.

**Comment by chrislovecnm** [comment]
I have asked Datastax to comment on this arch btw...

**Comment by bprashanth** [comment]
@chrislovecnm thanks for you input!

> this is different with vnodes - I am only recommending vnodes as best practice with kubernetes

I'm not sure we can get away with this given people will surely have older deployments that they want to get in Kube? 

> can we put the data on the k8 node and deploy a new docker? If so then you don't need to drain. If the data is going to be stored in the docker, then yes you need to drain
upgrade process depends on version jumps as well.

The data will be in a volume, not in the container. We will be unmounting the volume, updating docker, remounting the volume. But isn't it just safer to drain? 

I mean there will be a period of time the part of the ring you're responsible for will have no-one taking requests, and we need to be careful that this doesn't simultaneously hit all replicas responsible for that bit of the ring simultaneously, right? 

I was thinking something like: drain, rebalance ring, update drained node, readd to ring (approximately) 

> does k8 have a job scheduler on the roadmap? This could run nodetool repair.

Yeah cron scheduler would help, provided there is really no faiure scenario for nodetool repair that causes some fallout that needs non-cron like intervention: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/scheduledjob.md

> Bootstrapping, once you have the seeds up
Harder:
not sure what you mean

I mean it's easy to bootstrap as many nodes as required once we have the seeds up. This is different than some other forms of quorum databases where each node added to the cluster needs a delicate dance.

**Comment by chrislovecnm** [comment]
> I'm not sure we can get away with this given people will surely have older deployments that they want to get in Kube?

Sure if they have sstables that are using tokens then yes they would have to use them.  But frankly "best practices" probably should recommend using vnodes.  Adding capacity is much better handled by adding a few nodes, rather than doubling the size of the ring.  Also how are the token values being set?  You have to set the token value in `cassandra.yaml` or with `nodetool`.  You just don't have to rebalance the ring if you double the size, but you do have to set the token for each node.  Let me double check this, but I am fairly certain.

> But isn't it just safer to drain?

Kinda, but you are talking about theoretically moving TB of data around via network.  Just 100's of GB to drain a single node.  No need, and is inefficient. At times you have to shutdown the entire ring to upgrade, but that is another conversation :) Draining a node moves the data from that node onto another node.  Multiple nodes will stream the data to different nodes, but it is moving a lot of data around a network :) I usually would recommend to upgrade in place.  You drain a node when you want to remove one. 

>Yeah cron scheduler would help, provided there is really no faiure scenario for nodetool repair that causes some fallout that needs non-cron like intervention:

So a `nodetool repair` can fail, but if that happens, Houston we really have a problem.   So I assume that the job failure will have notifications.

I think of a seed as a regular node, all it is used for is when a new node boostraps.  The new node does not know the ring topography and has to lookup an ip to start chatting with.  Replacing it and removing them is a different process, which has a few specific steps.  I usually make the first three nodes in the ring seeds.   Because a seed node cannot be bootstrapped, you do not want every node to be like that.  But there is no steadfast rule.

I need to think about the seed provider more, and see if we can get some feedback from some gurus.

So a question for you... how does all of this roll up into PetSet?

**Comment by bprashanth** [comment]
> So a question for you... how does all of this roll up into PetSet?
 
Hand waving a little. 

Identity is defined as:
1. Stable network identity, i.e DNS and hostname
2. Stable storage identity, i.e a PV backed by a storage provider

PetSet will guarantee unique identities by this definition for each pet. The petset controller will start a pet, wait till it goes into running, then start the next. If a pet dies, another will be started with the same identity.  

This is the first cut. We will have to go out and implement a few pets, then iterate. 

For example, many pets have a bootstrapping sequence that would be easier to orchestrate from a babysitter running alongside the petset. The new pet will not go into ready till the babysitter has joined it to the cluster. In the case of Cassandra the babysitter could compute tokens/designate seeds (but there are alternate solutions as well). WDYT?

**Comment by bgrant0607** [comment]
cc @amygdala, since we'll eventually want to update the cassandra example

**Comment by chrislovecnm** [comment]
@bgrant0607 I am going to cut another ticket for the Snitch. @amygdala what is you C* background?!

**Comment by chrislovecnm** [comment]
@bprashanth 

> Stable storage identity, i.e a PV backed by a storage provider

So this is a no no ;) C* loves it some fast storage. Like local ssd. Dev and smaller deployments are ok with attached PV. Mission critical data is ok with PV, but make C* slow and cranky.

Does PetSet allow for no PV? Need to read the docs.

I am going to work on docs for the current snitch and get some opinions of this service discovery needs to be improved or is good to go!  I am sure it is fine, but would be good to double check.

**Comment by amygdala** [comment]
@chrislovecnm  - my C* background is near zilch :), look forward to consulting with you on best practices for this.

**Comment by chrislovecnm** [comment]
@amygdata appreciate the help, and hopefully some other folks that I have asked for help will have some time!

Also in regards to the storage stuff we are talking about for PetSet. This is another design thought that would apply:

https://github.com/kubernetes/kubernetes/issues/7562

**Comment by bprashanth** [comment]
> So this is a no no ;) C* loves it some fast storage. Like local ssd. Dev and smaller deployments are ok with attached PV. Mission critical data is ok with PV, but make C* slow and cranky.
Does PetSet allow for no PV? Need to read the docs.

This is slated for post 1.3. Once it works at the storage layer it should *just work* through petset. 

> I am going to work on docs for the current snitch and get some opinions of this service discovery needs to be improved or is good to go! 

Another thing that helps with Service discovery is init containers (https://github.com/kubernetes/kubernetes/pull/23666). It's a container that runs before your "actual" container and can, eg: declare its hostname into some shared storage, get the seed from the same shared storage and write its cassandra config file.

**Comment by chrislovecnm** [comment]
> Another thing that helps with Service discovery is init containers (#23666). It's a container that runs before your "actual" container and can, eg: declare its hostname into some shared storage, get the seed from the same shared storage and write its cassandra config file.

That might remove the need of a Custom Seed provider, but it does not address the need of a Snitch.  I am leaning towards a discovery service pattern.  This would allow for more dynamic capability, with possibly Snitch capability for C* racks and multiple C* data centers.

That is sorta the pattern that Mesos use, the build the Seed list at startup and then start the C* thread programmatically.  

I kinda like what C* has for EC2 and GCE discovery.

**Comment by bprashanth** [comment]
Honestly I haven't given the snitch that much though. What are its requirements? It needs to understand topology and which nodes are in which dc/rack etc right? Can we just write a custom kube-snitch that watches apiserver for endpoints? (relatively easy, I can help)

**Comment by chrislovecnm** [comment]
@bprashanth we are limited to single custer, without a rack capability at this point.  We have custom provider that watches the server endpoints.  See https://github.com/kubernetes/kubernetes/issues/24286

**Comment by bprashanth** [comment]
> The custom provider lists every node as a seed ... 

Yeah it's pretty bad, that's why I explicitly called out

> Perils:
Setting all nodes as seeds

above :)

> @bprashanth we are limited to single custer, without a rack capability at this point. We have custom provider that watches the server endpoints. See #24286

I think single cluster done *right* is a fine starting point. Of course the farther we get the better. Kubernetes is still figuring out cross cluster/AZ loadbalancing, we will be able to leverage what ever WAN   service discovery mechanism we settle on for that. 

> I would rather have a more flexible discovery service like what GCE or EC2 has in the C* source. Thoughts?

I haven't looked into those yet so I can't comment. I'm open to the idea and plan to investigate further once we get the basic cluster bootstrapping problems out of our way.

**Comment by chrislovecnm** [comment]
> once we get the basic cluster bootstrapping problems out of our way.

Yah lost me there.

**Comment by bprashanth** [comment]
I mean:
* Get a basic cassandra cluster working through petset (so not N Rcs and Services) without any admin fiddling. Of course this needs petset first for stable identities.
* Scale this cluster up/down safely (preferably without vnodes). This might require more custom c* code.
* Think about multi cluster. At this point we can consider targetting some kube resource from an autoscaler to say: deploy a second c* cluster to asia if the one in us-central is overloaded. 

Steps 1, 2 are actually harder today with containers than with vms, I'd like for that to not be the case.

**Comment by chrislovecnm** [comment]
@bprashanth without vnodes?  I have talked with multiple parties, and this may be an issue.  This is from teams running C* in production now.

And more C* code, let's try to not.

**Comment by bprashanth** [comment]
With vnodes is also a fine starting point. I'm willing to shave down the requirements for a V0 to as little as we can possibly get away with to have a fully functional c*. 

Can you please sketch out how we would:
1. Allocate seeds
2. Get tokens
3. Scale up/down
4. Add a new cluster in another AZ

With vnodes, provided we have petset, when you have the time?

**Comment by chrislovecnm** [comment]
1. Allocate seeds - have issue open already.
2. <s>Get tokens</s> - don't do that with vnodes.
3. Scale up/down - Scale up is working, scale down requires a trigger to drain and wait for the node to shutdown.  Run `nodetool decommission ( -h | --host ) <host name> | <ip address>` upon reducing the size of the cluster.  This is a looonnnnggg running process.
4. Add a new cluster in another AZ - this requires uber and or a better Snitch. This kinda corresponds to both the Seeds and the Snitch.  Seed need to be from multiple DC's for intance.

**Comment by bprashanth** [comment]
> Allocate seeds - have issue open already.

So from a quick look at the seed provider, I have a couple of questions. Looks like we're using endpoint ips https://github.com/chrislovecnm/kubernetes/blob/cassandra-seed-provider-docs/examples/cassandra/java/src/io/k8s/cassandra/KubernetesSeedProvider.java#L149 :
* What happens If the Service just doensn't have endpoints populated yet? (most people put the service def in the same file as the RC used for cassandra and `kubectl create -f`, so if the cassandra pods start running before the endpoints are populated all of them will set themselves as the seed right?) 
* What happens if the IP of your seed goes down and gets re-assigned to some other container (either a cassandra node in your cluster, another cassandra cluster or like an unrelated nginx pod) and we're bringing up a node that still observes the old list of endpoints (so it gets the wrong ip). 

Using the global endpoints list and IP doesn't sound reliable. I'd rather have each c* pod update, eg a config map with *its* own hostname (which it gets by running hostname), and takes the first member of the list in the config map as the seed. If we only start one at a time, it's a guaranteed single writer to that config map. Changes to the configmap are reflected in all pods that mount it. That hostname will never get re-assigned even when the pod gets rescheduled.

Doing that won't be easy without petset, but just throwing it out there to get your thoughts.

**Comment by chrislovecnm** [comment]
Appreciate the feedback.  I have an issue open here: https://github.com/kubernetes/kubernetes/issues/24286

Can we continue the conversation there?

**Comment by chrislovecnm** [comment]
@bprashanth I upated https://github.com/kubernetes/kubernetes/issues/24286 with your questions

**Comment by chrislovecnm** [comment]
@bprashanth  et all.  Good news the SeedProvider is a good design.  The reason that you don't want to have every node a seed is purely operational maint.  I will double check with a couple of other gurus, but initial review is good.

The SeedProvider only exists for initial bootstrap of a new node, and declaring the topography of a cluster.  The Snitch is the next fun item.

The main thing that I see is reducing the cluster size.  The k8s has to wait for the node to be drained.  Do we have an open issue for a k8s hook to run `nodetool drain` before a resize.

**Comment by chrislovecnm** [comment]
@bprashanth @bgrant0607 any idea when PetSet is due to arrive?

**Comment by chrislovecnm** [comment]
@bprashanth @bgrant0607 one other thing.  Will PetSet support exposing every Cassandra ip address and port to an external port?  A external client will need to see every C* server, not loadbalanced, but direct access.

**Comment by chrislovecnm** [comment]
@bgrant0607 @bprashanth any answer to the above question??

Also is there an issue in for a command to run before removing a pod?  For instance I reduce the size of a PetSet from 8 - 6.  In order to do that `nodetool decommission` needs to be run on the two nodes that are going down.

**Comment by bprashanth** [comment]
>any idea when PetSet is due to arrive?

Alpha in 1.3

> Will PetSet support exposing every Cassandra ip address and port to an external port? A external client will need to see every C* server, not loadbalanced, but direct access.

Not in alpha

> For instance I reduce the size of a PetSet from 8 - 6. In order to do that nodetool decommission needs to be run on the two nodes that are going down.

Is this only on scale-down, or do you need to run it when, eg I kubectl delete a node too? Can you just do it in a pre-stop hook or using deletionGracePeriodSeconds (http://kubernetes.io/docs/user-guide/production-pods/#lifecycle-hooks-and-termination-notice)? the petset will ensure that no other nodes are created before the decomm tool has done it's job and the pet is completely dead.
 
For alpha we're focussing on getting the bootstrapping patterns right for a variety of pets. I think it's safe to say there will be some form of event delivery for on-start and on-change, and if you really need to differentiate scale up/down we can possibly use the same mechanism.

**Comment by chrislovecnm** [comment]
@bprashanth do we have an issue in for properly exposing a Cassandra cluster?

**Comment by bprashanth** [comment]
Do you mean C* as a service? most of the people I've talked to (might be a biased sample) simply want c* in cluster for use in something like spark. What're the requirements, public DNS name for each node?

**Comment by chrislovecnm** [comment]
So for ubernetes, multiple datacenters, and client's that are external to k8s every cassandra pod (instance) ip / dns needs to be exposed.  This is the same for such systems as elasticsearch.

**Comment by chrislovecnm** [comment]
@bprashanth @bgrant0607 should I open this as a separate issue?

**Comment by bprashanth** [comment]
Kubernetes doesn't do external DNS yet. Once we get that cross platform, exposing it through petset should be no harder than exposing in-cluster DNS through petset. 

The petset will handle ordering, identity, cluster reconfiruation events and other common idioms required to safely deploy stateful services. Though we haven't clearly defined it's scope for WAN deployments, the petset controller probably *won't* go out and provision DNS itself, it will create another Kube resource that does so (an Ingress, federated Service etc).

**Comment by chrislovecnm** [comment]
@bprashanth completely understand, but do we have an open issue for that functionality?

**Comment by bprashanth** [comment]
External DNS has been disucssed standalone (https://github.com/kubernetes/kubernetes/issues/9914, https://github.com/kubernetes/kubernetes/issues/11760) in the context of Ingress (https://github.com/kubernetes/kubernetes/issues/24145) and in the context of Ubernetes (https://github.com/kubernetes/kubernetes/blob/8179a0d30e238c51a5c19e6217fe87633d40b1ae/docs/design/federation-phase-1.md).

**Comment by chrislovecnm** [comment]
@bprashanth https://github.com/kubernetes/kubernetes/pull/24912 makes me happy 👍

**Comment by chrislovecnm** [comment]
When will https://github.com/kubernetes/kubernetes/pull/24912 be in a tag??  kube-solo is awesome, but needs a tag 👍

**Comment by bprashanth** [comment]
The petset pr will surely be in the 1.3 release, it will probably make it into the next 1.3.0 alpha (https://github.com/kubernetes/kubernetes/releases). To declare petset alpha-feature-complete I need init containers and unready endpoints for services, though.

**Comment by chrislovecnm** [comment]
@bprashanth  @bgrant0607 any idea when when PetSet is alpha complete??  Would love it in the next alpha release ...

**Comment by bgrant0607** [comment]
@chrislovecnm suggested in #27074:

Move the example from https://github.com/k8s-for-greeks/gpmr/tree/master/pet-race-devops into the examples in this code base.

**Comment by chrislovecnm** [comment]
@bgrant0607 work in progress ;)

**Comment by jondubois** [comment]
@chrislovecnm Fully agree with your earlier suggestion about Kubernetes supporting some sort of user-defined hook/script which it could run before draining a node (e.g. when down-scaling a service). I think the utility value of this goes beyond Cassandra and would be useful for any stateful service. I have a similar problem with Redis - When down-scaling, I need some way to allow K8s to run a script on the target node which will tell my Redis cluster to remove the target node from the cluster (and give Redis Cluster the time it needs to rebalance its shards) and only once the rebalance operation is complete, K8s should be allowed to finish and remove the Pod/Container.

That sounds like it should be a separate issue? Is this similar to what you had in mind?

**Comment by chrislovecnm** [comment]
There is an issue open for sending signals HUP, etc, to pods. Don't have that handy since I am on my iPhone. Ping this ticket if you don't hear from me. The signal handling would allow us to script handlers to preform decomission.

Also we are looking at networking in multiple DC geos, without Federated / ubernetes. The last I knew that was not a good use case for redis, but that may have changed.

**Comment by chrislovecnm** [comment]
The key in my mind is to keep it simple and dumb. My mantral is to build toys that don't get me texted at 2 am ;)

**Comment by chrislovecnm** [comment]
Here: https://github.com/kubernetes/kubernetes/issues/27239#event-689568688 petsets in multiple Datacenters ;)

**Comment by chrislovecnm** [comment]
https://github.com/kubernetes/kubernetes/pull/26884

**Comment by chrislovecnm** [comment]
The pull above is signal handling. @github please give me edit on mobile ...

**Comment by mattbates** [comment]
@jondubois I know you asked about node-level hooks, but there appears to already be support for these types of hook at the _container_ level. See [PreStop hooks](http://kubernetes.io/docs/user-guide/container-environment/#container-hooks), with exec and HTTP handler implementations. Perhaps this might be worth exploring.

**Comment by jondubois** [comment]
@mattbates I didn't know about container hooks - That looks perfect for my use case. Thanks!

**Comment by bgrant0607** [comment]
cc @viglesiasce

**Comment by chrislovecnm** [comment]
@bgrant0607 is there a reason to keep this open?  I need to merge my Pet Set examples into examples.  But it is supported now.

**Comment by bprashanth** [comment]
Yeah upload a pr with "Fixes #24030" in the description :)

**Comment by bgrant0607** [comment]
@chrislovecnm Where is the example?

**Comment by paralin** [comment]
Can we start up another issue for Scylladb? It's the C implementation of Cassandra. I can write some snitches and stuff for it.

**Comment by chrislovecnm** [comment]
@paralin you probably want to look at submitting it to charts. Examples are going to start moving over to charts and actually be examples, not apps.

**Comment by paralin** [comment]
@chrislovecnm Thanks, will do.

**Comment by chrislovecnm** [comment]
@paralin to be clear https://github.com/kubernetes/charts - check the sig/apps agenda, we will be talking charts soon.

**Comment by pwittrock** [comment]
@chrislovecnm @bprashanth  I am moving this out of the blocking issues we are tracking for the 1.4 release.  If there is still 1.4 work to be done here, move it back and provide daily updates to this issue.

**Comment by chrislovecnm** [comment]
@pwittrock this is solid to go into 1.4, but I don't want to impede anything as well. Don't know quite how your process is handled when you are close to a release.

**Comment by chrislovecnm** [comment]
@pwittrock let me add, my PR for C* and PetSet got a LGTM today. It is listed above, we should be able to close this now. I am going to move on to getting a chart done, and then we can access if we want to pull it out of examples and into only charts.

**Comment by pwittrock** [comment]
@chrislovecnm I now have to focus on issues that are putting the release at real risk and cannot focus on anything non-critical to shipping 1.4 in just over 1 week.

**Comment by chrislovecnm** [comment]
@pwittrock hence the comment about not wanting to impede anything ...  Yay, release time, pass me the caffeine please.

**Comment by chrislovecnm** [comment]
@pwittrock https://github.com/kubernetes/kubernetes/pull/30577 is in, shall we close?

**Comment by pwittrock** [comment]
SGTM
